app/javascript/mastodon/features/ui/components/navigation_bar.tsx

```diff
@@ -0,0 +1,217 @@
+import { useCallback, useEffect } from 'react';
+
+import { useIntl, defineMessages, FormattedMessage } from 'react-intl';
+
+import classNames from 'classnames';
+import { NavLink, useRouteMatch } from 'react-router-dom';
+
+import AddIcon from '@/material-icons/400-24px/add.svg?react';
+import HomeActiveIcon from '@/material-icons/400-24px/home-fill.svg?react';
+import HomeIcon from '@/material-icons/400-24px/home.svg?react';
+import MenuIcon from '@/material-icons/400-24px/menu.svg?react';
+import NotificationsActiveIcon from '@/material-icons/400-24px/notifications-fill.svg?react';
+import NotificationsIcon from '@/material-icons/400-24px/notifications.svg?react';
+import SearchIcon from '@/material-icons/400-24px/search.svg?react';
+import { openModal } from 'mastodon/actions/modal';
+import { toggleNavigation } from 'mastodon/actions/navigation';
+import { fetchServer } from 'mastodon/actions/server';
+import { Icon } from 'mastodon/components/icon';
+import type { IconProp } from 'mastodon/components/icon';
+import { IconWithBadge } from 'mastodon/components/icon_with_badge';
+import { useIdentity } from 'mastodon/identity_context';
+import { registrationsOpen, sso_redirect } from 'mastodon/initial_state';
+import { selectUnreadNotificationGroupsCount } from 'mastodon/selectors/notifications';
+import { useAppDispatch, useAppSelector } from 'mastodon/store';
+
+const messages = defineMessages({
+  home: { id: 'tabs_bar.home', defaultMessage: 'Home' },
+  search: { id: 'tabs_bar.search', defaultMessage: 'Search' },
+  publish: { id: 'tabs_bar.publish', defaultMessage: 'New Post' },
+  notifications: {
+    id: 'tabs_bar.notifications',
+    defaultMessage: 'Notifications',
+  },
+  menu: { id: 'tabs_bar.menu', defaultMessage: 'Menu' },
+});
+
+const IconLabelButton: React.FC<{
+  to: string;
+  icon?: React.ReactNode;
+  iconComponent?: IconProp;
+  activeIcon?: React.ReactNode;
+  activeIconComponent?: IconProp;
+  title: string;
+}> = ({ to, icon, iconComponent, activeIcon, activeIconComponent, title }) => {
+  const match = useRouteMatch(to);
+
+  return (
+    <NavLink
+      className='ui__navigation-bar__item'
+      activeClassName='active'
+      to={to}
+      aria-label={title}
+    >
+      {match?.isExact && activeIcon ? activeIcon : icon}
+      {iconComponent && (
+        <Icon
+          id=''
+          icon={
+            match?.isExact && activeIconComponent
+              ? activeIconComponent
+              : iconComponent
+          }
+        />
+      )}
+    </NavLink>
```

Comment:
I found this bit of code hard to wrap my head around as just going by the internal logic it's entirely possible for two icons to be rendered in one link which seems unexpected. Might be simpler to forego the `iconComponent`/`activeIconComponent` shortcut and always passing in `ReactNodes`